Implement frequency-differencing mask

b-reyes opened the pull request:
This PR addresses issue #224 by constructing a function that computes a frequency-differencing mask for a given `Sv` data source. Please note that this function is placed within the `mask` sub-package as determined by https://github.com/OSOceanAcoustics/echopype/pull/817#issuecomment-1294332886.

b-reyes commented on echopype/utils/data_proc_lvls.py:
```diff
@@ -0,0 +1,66 @@
+from echopype.utils.io import check_file_existence, get_file_format, validate_output_path
+
+
+def validate_source_Sv(
```
I just realized that there is no part of this function that requires that the input correspond to `Sv` data. So we can use this in other functions (in particular `apply_mask`) I suggest we remove any mention of `Sv` and replace it with `ds`.

leewujung replied:
I agree with this.

Also I think we should move this to `utils.io` instead of having a separate module for this, especially since the functions used are from `utils.io` already. If there are needs in the future to have many of this type of functions, we can split them out to a separate module under `utils`.

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+            "The Dataset defined by source_Sv must have frequency_nominal as a variable!"
+        )
+
+    # make sure that the channel and frequency_nominal values are not repeated in source_Sv
+    if len(set(source_Sv.channel.values)) != 2:
+        raise RuntimeError(
+            "The provided source_Sv contains repeated channel values, " "this is not allowed!"
+        )
+
+    if len(set(source_Sv.frequency_nominal.values)) != 2:
+        raise RuntimeError(
+            "The provided source_Sv contains repeated frequency_nominal "
+            "values, this is not allowed!"
+        )
```
`source_Sv` should be able to have more than 2 channels (and users select only 2 of those to create the mask). Shouldn't the check be something like
```
len(set(source_Sv.channel.values)) < source_Sv.channel.size
```
The same for `frequency_nominal`

b-reyes replied:
Yes, you are correct. Thank you for catching that and providing a solution!

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+            "values, this is not allowed!"
+        )
+
+    def get_positions(var_name, input_list):
+
+        # obtain position of input in either frequency_nominal or channel
+        inputA_pos = np.argwhere(source_Sv[var_name].values == input_list[0]).flatten()
+        inputB_pos = np.argwhere(source_Sv[var_name].values == input_list[1]).flatten()
+
+        return inputA_pos, inputB_pos
+
+    # get the position of freqAB or chanAB values in frequency_nominal or channel, respectively
+    if freqAB is not None:
+        inputA_pos, inputB_pos = get_positions(var_name="frequency_nominal", input_list=freqAB)
+    else:
+        inputA_pos, inputB_pos = get_positions(var_name="channel", input_list=chanAB)
+
+    # check that the elements of freqAB or chanAB are in frequency_nominal or channel, respectively
+    if (len(inputA_pos) == 0) or (len(inputB_pos) == 0):
+
+        if freqAB is not None:
+            raise RuntimeError(
+                "The provided list input freqAB contains values that "
+                "are not in the frequency_nominal variable!"
+            )
+        else:
+            raise RuntimeError(
+                "The provided list input chanAB contains values that are "
+                "not in the channel coordinate!"
+            )
```
Why do you need to determine the position to know whether the elements of `freqAB` or `chanAB` are in the dataset? Wouldn't that just take a list comprehension like
```
np.all([f in souce_Sv["frequency_nominal] for f in freqAB])
```
and equivalent for `channel` and raise RuntimeError when the above is `False`?

b-reyes replied:
You are right, there is a better way to implement this using list comprehension. I previously was utilizing `get_positions` such that this made sense, but I ended up removing it. I will go ahead and improve this using your suggestions.

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+        A Dataset containing the Sv data
+    """
+
+    source_Sv, file_type = validate_source_Sv(source_Sv, storage_options)
+
+    if isinstance(source_Sv, str):
+
+        # open up Dataset using source_Sv path
+        source_Sv = xr.open_dataset(source_Sv, engine=file_type, chunks="auto", **storage_options)
```
I think that these few lines should be moved outside of this function and into the main block of `frequency_difference`, and that way you can have this function `_check_source_Sv_freq_diff` really just really handles the checks, instead of returning a Dataset that is somewhat awkward and incompatible with the function name.

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+    # create mask using operator lookup table
+    da = xr.where(str2ops[operator](lhs, diff), True, False)
+
+    # assign a name to DataArray
+    da.name = "frequency_difference_mask"
```
I think we can just call this mask right now and leave the other parts to the provenance that should be recorded in attributes (i.e., this mask is created using freq diff function).
```suggestion
    # assign a name to DataArray
    da.name = "mask"
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+    return source_Sv
+
+
+def frequency_difference(
```
```suggestion
def frequency_differencing(
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+    storage_options: dict, optional
+        Any additional parameters for the storage backend, corresponding to the
+        path provided for ``source_Sv``
+    freqAB: list of float, optional
+        The pair of nominal frequencies to be used for frequency-differencing, where
+        the first element corresponds to ``freqA`` and the second element corresponds
+        to ``freqB`` (see notes below for more details)
+    chanAB: list of float, optional
+        The pair of channels that will be used to select the nominal frequencies to be
+        used for frequency-differencing, where the first element corresponds to ``freqA``
+        and the second element corresponds to ``freqB`` (see notes below for more details)
```
```suggestion
    freqAB: list of float, optional
        The pair of nominal frequencies to be used for frequency-differencing, where
        the first element corresponds to ``freqA`` and the second element corresponds
        to ``freqB``. Only one of ``freqAB`` and ``chanAB`` should be provided, and not both.
    chanAB: list of float, optional
        The pair of channels that will be used to select the nominal frequencies to be
        used for frequency-differencing, where the first element corresponds to ``freqA``
        and the second element corresponds to ``freqB``.
        Only one of ``freqAB`` and ``chanAB`` should be provided, and not both.
```

leewujung commented on echopype/mask/api.py:
```diff
@@ -0,0 +1,306 @@
+
+    Notes
+    -----
+    Either ``freqAB`` or ``chanAB`` must be provided, but both parameters
+    cannot be given.
```
```suggestion
```
I suggest removing these and added description upfront under both argin description so that it is read in one place.

leewujung commented on echopype/tests/mask/test_mask.py:
```diff
@@ -0,0 +1,222 @@
+from typing import List, Union
+
+
+def get_mock_freq_diff_data(n: int, add_chan: bool, add_freq_nom: bool):
```
Because of [my comment above](https://github.com/OSOceanAcoustics/echopype/pull/901/files#r1049171963), I think this function should allow creating a dataset with variable number of channels/frequencies, and add a couple tests for selecting two of the X channels/frequencies out of the many channels/frequencies to make sure things run through.

b-reyes replied:
That is a good call! I will go ahead and add this functionality.

leewujung commented on echopype/utils/data_proc_lvls.py:
```diff
@@ -0,0 +1,66 @@
+    source_Sv: xr.Dataset or str
+        A Dataset that contains the Sv data (which will be the same as
+        the input) or a validated path to a zarr or netcdf file
+    file_type: {"netcdf4", "zarr"}, optional
+        If ``source_Sv`` is a path then corresponds to the file type of input, else
+        is ``None``
```
```suggestion
    file_type: {"netcdf4", "zarr"}, optional
        The file type of the input path if ``source_Sv`` is a path, otherwise ``None``
```

leewujung reviewed (changes requested):
@b-reyes : thanks for submitting this. I made a number of suggestions. I think there's a bug in the checking of duplicated channels/frequency, and it'll also be good to have a test for making sure the selection (selecting 2 out of X channels/frequencies) goes through correctly.

b-reyes commented:
> @b-reyes : thanks for submitting this. I made a number of suggestions. I think there's a bug in the checking of duplicated channels/frequency, and it'll also be good to have a test for making sure the selection (selecting 2 out of X channels/frequencies) goes through correctly.

@leewujung I have implemented your suggestions, corrected the bugs, and extended the mock Dataset and associated tests so that you can select 2 out of X channels/frequencies. This PR is ready for another review.

leewujung commented on echopype/tests/utils/test_utils_io.py:
```diff
@@ -258,3 +259,43 @@ def test_env_indep_joinpath_os_dependent(save_path: str, is_windows: bool, is_cl
+)
+def test_validate_source_ds(source_ds_input, storage_options_input, true_file_type):
+    """
+    Tests that ``validate_source_ds`` has the appropriate outputs. It should
+    be noted that an exhaustive combination of ``source_ds`` and ``storage_options``
+    has not been implemented when ``source_ds`` is a path. These are tested in
+    ``test_validate_output_path``.
```
Please check if what I modified in the below is correct:
```suggestion
    Tests that ``validate_source_ds`` has the appropriate outputs. 
    An exhaustive list of combinations of ``source_ds`` and ``storage_options``
    are tested in ``test_validate_output_path`` and are therefore not included here.
```

leewujung reviewed (approved):
@b-reyes : the changes look good! I just have a suggested comment change for `test_validate_source_ds`. Once you have a look and verify that it is correct or change if it is not, please go ahead and merge this.

b-reyes commented:
> @b-reyes : the changes look good! I just have a suggested comment change for test_validate_source_ds. Once you have a look and verify that it is correct or change if it is not, please go ahead and merge this.

I took a look at your comment change for `test_validate_source_ds` and it looks good. I have commited it. I will go ahead and merge this once the tests pass again.
